### DISCUSSION TITLE
Introduce type safety to CatchBoundary?

### DISCUSSION THREAD

**Discussion by binajmen** [remix-run/remix#5496] [Proposals]
Hi 👋 

I was reflecting on what could be the right approach when you need to handle multiple cases in the CatchBoundary of one route.

My first thought was to define a 
```ts
const enum Exception {
  NotAllowed,
  NoData,
  BadFilter
}
```

to pass for example `Exception.BadFilter` as the message in the `Response` I throw:
```ts
throw new Response(Exception.BadFilter, { status: 400 })
```

I could then use a switch in the CatchBoundary to render the right info:
```ts
export function CatchBoundary() {
  const caught = useCatch();

  switch (caught.data) {
    case Expection.BadFilter:
      return <BadFilter />
    ...
  }
}
```
_(this is pseudo code, you get the idea)_

It could work, but it means I can't really embed satellite information: in the case of a bad filter, I might want to show the filter that was provided but rejected.

One way would be to `JSON.stringify({ exception: Expection.BadFilter, filter: currentFilter })` for the message, then use `zod` in the `CatchBoundary` to parse (with `discriminatedUnion`) the message's `Response`:
```ts
const catchSchema = z.discriminatedUnion("exception", [
  z.object({ exception: z.literal(Expection.BadFilter), filter: z.string() }),
  ...
]);
```

But I didn't think about the fact `CatchBoundary` is run on the client, so zod itself will be shipped to the client, and we all know it's not good (for bundle size reason).

@tom-sherman shared in a Discord discussion ([read the discussion here](https://discord.com/channels/770287896669978684/1076048659357114418/1076048659357114418)) how he (would?) handle complex cases by avoiding using `CatchBoundary` and rather doing:
```ts
export default function SomeRouteSwitch() {
  const data = useLoaderData<typeof loader>()

  if (data.ok === true) return <SomeRoute />
  if (data.ok === false) return <CatchBoundary />
}

function SomeRoute() {}

function CatchBoundary() {}

// Export this wherever appropriate
export function ErrorBoundary() {}
```
Why? Because "Throwing things is inherently not type safe", and I agree with that. By returning instead of throwing, we can have type safety. But in a way, it defeats the purpose of CatchBoundary (and is kind of boilerplate-y).

A first suggestion drafted in the discussion was:
```ts
import { ThrownResponse, useCatch } from '@remix-run/react';

export function createThrownResponseHelper<T>(): {
  useCatch: () => ThrownResponse<number, T>;
  errorResponse: (body: T, init?: ResponseInit) => Response;
} {
  return {
    useCatch: () => {
      const caught = useCatch();

      return {
        ...caught,
        data: JSON.parse(caught.data),
      };
    },
    errorResponse: (body, init) => {
      const headers = new Headers(init?.headers);
      headers.set('Content-Type', 'appplication/json');
      return new Response(JSON.stringify(body), {
        ...init,
        headers,
      });
    },
  };
}

const { useCatch, errorResponse } =
  createThrownResponseHelper<{ foo: number }>();

export function loader() {
  // This is a type error
  // throw errorResponse({ foo: "hello" });

  // This guarantees we return the correct shape
  throw errorResponse({ foo: 4 });
}

export function CatchBoundary() {
  const caught = useCatch();
  console.log(caught.data.foo); // 4
}
```

Again, not a final proposition, more a food for thought.

I would be interested to get your feedback on the subject, and see how you handle this :)

**Comment by kentcdodds** [comment]
Any reason you're not just doing `throw json({whatever: 'stuff'}, {status: 400})`
